pulp/apis/sas_api.py

```diff
@@ -0,0 +1,657 @@
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+# SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE."""
+
+from .core import LpSolver_CMD, LpSolver, PulpSolverError, log
+from io import StringIO
+from contextlib import redirect_stdout
+import os
+import sys
+from .. import constants
+import warnings
+from typing import Union
+
+from uuid import uuid4
+
+# The maximum length of the names of variables and constraints.
+MAX_NAME_LENGTH = 256
+
+# This combines all status codes from OPTLP/solvelp and OPTMILP/solvemilp
+SOLSTATUS_TO_STATUS = {
+    "OPTIMAL": constants.LpStatusOptimal,
+    "OPTIMAL_AGAP": constants.LpStatusOptimal,
+    "OPTIMAL_RGAP": constants.LpStatusOptimal,
+    "OPTIMAL_COND": constants.LpStatusOptimal,
+    "TARGET": constants.LpStatusOptimal,
+    "CONDITIONAL_OPTIMAL": constants.LpStatusOptimal,
+    "FEASIBLE": constants.LpStatusNotSolved,
+    "INFEASIBLE": constants.LpStatusInfeasible,
+    "UNBOUNDED": constants.LpStatusUnbounded,
+    "INFEASIBLE_OR_UNBOUNDED": constants.LpStatusUnbounded,
+    "SOLUTION_LIM": constants.LpStatusNotSolved,
+    "NODE_LIM_SOL": constants.LpStatusNotSolved,
+    "NODE_LIM_NOSOL": constants.LpStatusNotSolved,
+    "ITERATION_LIMIT_REACHED": constants.LpStatusNotSolved,
+    "TIME_LIM_SOL": constants.LpStatusNotSolved,
+    "TIME_LIM_NOSOL": constants.LpStatusNotSolved,
+    "TIME_LIMIT_REACHED": constants.LpStatusNotSolved,
+    "ABORTED": constants.LpStatusNotSolved,
+    "ABORT_SOL": constants.LpStatusNotSolved,
+    "ABORT_NOSOL": constants.LpStatusNotSolved,
+    "OUTMEM_SOL": constants.LpStatusNotSolved,
+    "OUTMEM_NOSOL": constants.LpStatusNotSolved,
+    "FAILED": constants.LpStatusNotSolved,
+    "FAIL_SOL": constants.LpStatusNotSolved,
+    "FAIL_NOSOL": constants.LpStatusNotSolved,
+    "ERROR": constants.LpStatusNotSolved,
+}
+
+
+class SASsolver(LpSolver_CMD):
+    name = "SASsolver"
+
+    def __init__(
+        self,
+        mip=True,
+        msg=True,
+        warmStart=False,
+        keepFiles=False,
+        timeLimit=None,
+        **solverParams,
+    ):
+        """
+        :param bool mip: if False, assume LP even if integer variables
+        :param bool msg: if False, no log is shown
+        :param bool warmStart: if False, no warm start
+        :param bool keepFiles: if False, the generated mps mst files will be removed
+        :param solverParams: SAS proc OPTMILP or OPTLP parameters
+        """
+        LpSolver_CMD.__init__(
+            self, mip=mip, msg=msg, warmStart=warmStart, keepFiles=keepFiles
+        )
+        self.timeLimit = timeLimit
+        # Only named options are allowed in SAS solvers.
+        self.solverOptions = {key: value for key, value in solverParams.items()}
+
+    def _create_statement_str(self, statement):
+        """Helper function to create the strings for the statements of the proc OPTLP/OPTMILP code."""
+        stmt = self.solverOptions.pop(statement, None)
```

Comment:
please do not edit this property, it should be static. There are many places where it's edited in the code.